analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.cs

```diff
@@ -300,4 +300,22 @@ public void Foo()
             MyContainer.publicStaticDictionaryField.Add("x", "x1"); // Noncompliant
         }
     }
+
+    // https://github.com/SonarSource/sonar-dotnet/issues/4178
+    public class Repro
+    {
+        public void DifferentObjectSameProperty()
+        {
+            var first = new ClassWithListProperty();
+            var second = new ClassWithListProperty();
+
+            first.IntList[0] = 1;
+            second.IntList[0] = 2;
+        }
+
+        private class ClassWithListProperty
+        {
+            public List<int> IntList { get; set; }
```

Comment:
could you also add IntList2, and then the test above should be:

```
            first.IntList[0] = 1;
            first.IntList2[0] = 11;
            second.IntList[0] = 2;
            second.IntList2[0] = 22;
```
and no issue should be raised

Also, could you also add separate tests with various combinations of correctly raised issues like

```
            first.IntList[0] = 1;
            first.IntList2[0] = 2;
            first.IntList[0] = 3; // Noncompliant
            first.IntList2[1] = 4; // ok, different index
```

and 

```
            first.IntList[0] = 1;
            first.IntList2[1] = 2;
            first.IntList[1] = 3; 
            first.IntList2[1] = 4; // Noncompliant
```

and more compliant test cases like:
```
            first.IntList[0] = 1;
            first.IntList2[1] = 2;
            first.IntList[1] = 3; 
            first.IntList2[0] = 4;
```

Reply:
also, can you add some tests with nested objects, like

```
private class ClassWithNested
{
            public ClassWithListProperty Alpha { get; set; }
            public ClassWithListProperty Beta { get; set; }
}
```

and something like

```
            first.Alpha.IntList[0] = 1;
            first.Alpha.IntList2[0] = 11;
            first.Beta.IntList[0] = 1111;
            first.Beta.IntList2[0] = 11111;
            second.Alpha.IntList[0] = 2;
             second.Alpha.IntList2[0] = 22;
            second.Beta.IntList[0] = 222;
            second.Beta.IntList2[0] = 2222;
```

and a noncompliant case with such nested structure